### PR TITLE
Typedef'ed layer_state_t to uint32_t.

### PR DESCRIPTION
**Pull request by alex-ong**
This lets us change layer_state_t to uint8_t in the case that there are 8 layers or less, saving bytes and processing time. 

One of the major bottlenecks of key processing is iterating through all layers to see which keycode to process. This commit opens up the future possibility for:

- Determining NUM_LAYERS 
- type-defing layer_state_t to less bytes for less storage, and faster processing
- Counting down from NUM_LAYERS rather than 31

See similar pull request for TMK, which seems to have support not merged yet.
https://github.com/tmk/tmk_keyboard/pull/526

### CODE REVIEW

**Comment by drashna** [comment]
There are a bunch of merge conflicts here now.  Would you mind fixing this?

**Comment by alex-ong** [comment]
Merge conflicts fixed! I havent tested this on a real keyboard, on account that i don't have one handy, but it was working before the i fixed the merge conflicts...

**Comment by drashna** [comment]
Could you fix the merge conflicts?

**Comment by alex-ong** [comment]
> Could you fix the merge conflicts?

Fixed!

**Comment by drashna** [comment]
I hate to do this to you, but there are a number of layer functions in quantum.[ch].  Would you mind checking them, at least, and update any of them that need to be updated? 

I don't think they need to be, based on how they work, but ... just in case. 

Also, there are some docs that should be updated. 
https://github.com/qmk/qmk_firmware/blob/master/docs/custom_quantum_functions.md#layer-change-code
https://docs.qmk.fm/#/custom_quantum_functions?id=example-implementation
https://github.com/qmk/qmk_firmware/blob/master/docs/ref_functions.md#example-1

These use the layer codes, and probably should be updated.

**Comment by alex-ong** [comment]
Ahh, i just realised that people use uint32_t everywhere in their keyboard code - should i be updating that (and importing action_layer.h) where necessary or leaving uint32_t everywhere?

**Comment by drashna** [comment]
Honestly, I'm not sure.  
I think adding it to the documentation should be okay.  Right now, it looks like it should be fine, if it's left as is.  So it still compiles, and doesn't affect anyone's code. 

But let me tag @skullydazed and @jackhumbert about that, to see what we'd want to do.  But I think it should be fine as is. 


Also, as it stands, there is no way to change this, without changing `action_layer.h`, is there? 
I can think of a couple of boards that could use uint8_t instead, already (VIA only supports 4 layers, due to limitations in EEPROM size, so ... we don't need to use uint32_t for those boards).

File: tmk_core/common/action_layer.h [hunk 1]
```diff
@@ -21,24 +21,24 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "keyboard.h"
 #include "action.h"
 
-
+typedef uint32_t layer_state_t;
```
Comment (drashna) [inline tmk_core/common/action_layer.h]:
```suggestion
#if defined(LAYER_STATE_8BIT) || defined(DYNAMIC_KEYMAP_ENABLE)
typedef uint8_t layer_state_t;
#elif defined(LAYER_STATE_16BIT)
typedef uint16_t layer_state_t;
#else
typedef uint32_t layer_state_t;
#endif
```

Comment (drashna) [reply tmk_core/common/action_layer.h]: The `DYNAMIC_KEYMAP_ENABLE` option is for ... well VIA mostly, which uses EEPROM for keymap store the layers, and is limited to 4 usually.

Comment (drashna) [reply tmk_core/common/action_layer.h]: @alex-ong Any thoughts or updates on this?

Comment (alex-ong) [reply tmk_core/common/action_layer.h]: sorry, been busy with work... hopefully i can get these changes done. 
From the TMK pr i remember that although `typedef uint16_t layer_state_t` compiles, its also good to change it in `action.c ` (refer to https://github.com/tmk/tmk_keyboard/pull/526)
Thoughts on that?

**Comment by drashna** [comment]
[Typedef'ing Layer variables](https://trello.com/c/rZRzfUFN/4-typedefing-layer-variables)

File: tmk_core/common/action_layer.h (hunk 1 above)
Comment (drashna) [reply tmk_core/common/action_layer.h]: Not a problem. Work comes first! 

That would be best, definitely.  

However, I think a define is fine for now.   Both the Ergodox EZ and QMK Configurators could check the number of layers and add the define at compile time, until such a time that is possible/added.

**Comment by drashna** [comment]
The quantum.h file needs to be update too:
https://github.com/qmk/qmk_firmware/blob/master/quantum/quantum.h#L64-L68

**Comment by drashna** [comment]
Also: default_layer_state_set_user needs to be changed.

**Comment by drashna** [comment]
@alex-ong Have you had a chance to look over the suggestions?

**Comment by alex-ong** [comment]
I merged your changes.
Only thing to consider is whether this should also be packed in, since shifting will break unless we have this?
```
#if NUM_LAYERS > 16   // layer_state is unit32_t
    uint8_t shift = action.layer_bitop.part*4;
#elif NUM_LAYERS > 8   // unit16_t
    uint8_t shift = (action.layer_bitop.part & 3)*4;
#else  // uint8_t
    uint8_t shift = (action.layer_bitop.part & 1)*4;
#endif
```

or in this case, reversed...
```
#if defined(LAYER_STATE_8BIT) || ( defined(DYNAMIC_KEYMAP_ENABLE) && DYNAMIC_KEYMAP_LAYER_COUNT <= 8 )  // uint8_t
    uint8_t shift = (action.layer_bitop.part & 1)*4;
#elif defined(LAYER_STATE_16BIT) > 8   // uint16_t
    uint8_t shift = (action.layer_bitop.part & 3)*4;
#else   // layer_state is uint32_t
    uint8_t shift = action.layer_bitop.part*4;
#endif
```

File: tmk_core/common/action_layer.h (hunk 1 above)
Comment (alex-ong) [reply tmk_core/common/action_layer.h]: > ```#if defined(LAYER_STATE_8BIT) || ( defined(DYNAMIC_KEYMAP_ENABLE) && DYNAMIC_KEYMAP_LAYER_COUNT >= 8 )```
Just realised is it meant to be <= 8, or >= 8?

**Comment by drashna** [comment]
Ah, yeah, that would be a problem. 

biton32 may be a problem too:
https://github.com/qmk/qmk_firmware/blob/master/tmk_core/common/util.c#L51-L79

Though, I think that may be okay?

**Comment by alex-ong** [comment]
> Ah, yeah, that would be a problem.
> 
> biton32 may be a problem too:
> https://github.com/qmk/qmk_firmware/blob/master/tmk_core/common/util.c#L51-L79
> 
> Though, I think that may be okay?

biton32 itself is fine, the problem is that every keyboard calls it to find out which layer they are on...
`uint8_t activeLayer = biton32(layer_state)`

ideally this is replaced with 
`uint8_t activeLayer = getActiveLayer(layer_state)`

```
#if 8_bits
uint8_t getActiveLayer(layer_state_t state)
{
    biton8(state)
}
```

**Comment by drashna** [comment]
Thanks!

**Comment by jceb** [comment]
This commit breaks layer switching on my signum 3.0. Anyone else experiencing issues with it?

**Comment by noroadsleft** [comment]
Confirmed. Flashed my existing keymap from a current master and couldn't switch layers. Reverted this commit locally and reflashed, and layer switching works again.

File: tmk_core/common/action_layer.c [hunk 2]
```diff
@@ -301,9 +301,9 @@ uint8_t layer_switch_get_layer(keypos_t key) {
   action_t action;
   action.code = ACTION_TRANSPARENT;
 
-  uint32_t layers = layer_state | default_layer_state;
+  layer_state_t layers = layer_state | default_layer_state;
   /* check top layer first */
-  for (int8_t i = 31; i >= 0; i--) {
+  for (int8_t i = sizeof(layer_state_t)-1; i >= 0; i--) {
```
Comment (XScorpion2) [inline tmk_core/common/action_layer.c]: This is probably the issue. `sizeof(layer_state_t)-1` returns 3, not 31 as this is the size in bytes of the type. This should be  `sizeof(layer_state_t) * 8 - 1`

**Comment by drashna** [comment]
Odd. I have a new commit that builds on these changes and may address those issues

**Comment by XScorpion2** [comment]
@drashna https://github.com/qmk/qmk_firmware/pull/5906 does not fix this issue.

File: tmk_core/common/action_layer.c (hunk 2 above)
Comment (noroadsleft) [reply tmk_core/common/action_layer.c]: Can confirm changing this as described fixes layer switching.

**Pull request merged**